Silex 2

kordos opened the pull request:
Hi,

This changes allow to use silex2 with php di.
Changes:
1. ArgumentResolver class: silex 2 uses it for resolving controller arguments so I moved this code from ControlerResolver to new class.

2. Proxy for request in Application - now php di is able to inject real Request object which Application uses internally. 
Example definition for request class:
Request::class => \DI\get('request'),

In 1.x such DI definition allows to inject Request class used in Application and now also in 2.x:

Commit:
HttpKernelServiceProvider - added access to request plus tests, added composer.lock

Commit:
added ArgumentResolver class with implementation taken from ControllerResolver

mnapoli commented on .gitignore:
```diff
@@ -1,3 +1,3 @@
 /vendor/
 /composer.phar
-/composer.lock
```
Please keep this line and remove `composer.lock` from the pull request.

kordos replied:
I can remove it but why you don't want to have `composer.lock` file in repository? Thanks to `composer.lock` everyone who will be working on that library will have the same version of dependencies.

jdreesen replied:
Because the `composer.lock` file is not ment do be committed to the repository for libraries.

It also causes Travis to fail, because Travis tests this library with various PHP versions but the `composer.lock` pins the dependencies to specific versions that are not compatible with older PHP versions.

kordos replied:
ok, In such case I will remove composer.lock file

mnapoli commented on src/Provider/HttpKernelServiceProvider.php:
```diff
@@ -61,6 +58,21 @@ public function register(\Pimple\Container $app)
+        // add request to ensure that DI have access to correct request instance from application
+        $app['request'] = $app->factory(function ($app) {
+            return $app['request_stack']->getCurrentRequest();
+        });
```
Since the container should be stateless (and all the registered services too) the request should not be stored in the container. Instead services should inject the request stack and resolve the request when they need it.

Here the request will be resolved once and stored by the container (so the same object will be returned every time).

kordos replied:
Request is assign by factory method so `request` object will be instance which currently is used in Application. Every call will return current request instance.

In few places in application I'm using `Request` object. Without that I'm getting new instance and should be instance used by application with filled arguments. I don't want to change it to `RequestStack` in every occurrence. So this solution seams valid: if you have `Request` object to inject in other places than controller you want to receive original `Request` object not new one.

mnapoli replied:
Ah you're right I thought it was a "PHP-DI factory", which is not resolved every time.

Anyway there should be only stateless objects in the container ([here is a good intro](https://igor.io/2013/03/31/stateless-services.html) on the subject). You can register that binding in your projects though if you want it that way.

kordos replied:
Ok, I can remove it if you don't want to have it here. I will register that binding in my project.

mnapoli commented on src/Controller/ControllerResolver.php:
```diff
@@ -61,41 +53,10 @@ public function getController(Request $request)
 
     /**
      * {@inheritdoc}
+     * @deprecated This method is deprecated. Silex 2.x uses ArgumentResolverInterface for controller arguments
```
Why keep this method and implement `ArgumentResolverInterface` on this class? Could it be removed completely?

kordos replied:
`ControllerResolverInterface` still have this method with comment that it is deprecated a will be removed in 4.0 version so it have to stay until then.

jdreesen replied:
I think the implementation should stay too, because we need to support Symfony < 3.1 (because Silex itself supports 2.8+) which does not have a `ArgumentResolverInterface` and still uses this method.

kordos replied:
I focused on newer version. If you want to support symfony 2.8+ then yes. But then also HttpKernel class have to check that version and depending on it assigning right implementation.

jdreesen replied:
Well, as this is a bridge between PHP-DI and Silex, I think it makes sense to support the full dependency range of both libraries (unless there are good reasons not to). And Silex (even in its latest version) supports Symfony 2.8+ (because Symfony 2.8 is an LTS version with support until November 2019).

@mnapoli What's your opinion on this?

mnapoli replied:
Yes good point, if it's possible to keep the compatibility let's keep it.

kordos replied:
Ok, I will check what is needed here and try to add support also for 2.8 version

mnapoli reviewed (commented):
Thank you for doing this work!

I have added a few comments or questions inline.

mnapoli commented:
The build is failing because of PHP versions it seems. Which PHP version is required by Silex 2?

We could bump the requirements to these versions. Feel free also to drop HVVM from Travis (I'm dropping support on all my projects).

mnapoli commented:
One last comment :) : you should update `composer.json` to require Silex v2

kordos commented:
There is already silex 2 in composer.json.

About PHP version:
silex 2 requires php >=5.5.9 and I'm using php 7.1
Without composer.lock build will be probably working correctly. Please answer to my comment about it.

Commit:
Removed composer.lock file and added to .gitignore

kordos commented:
Now build is failing for php 5.5.x version which isn't compatible with silex 2. Silex to requires php in >=5.5.9.
Other test is with HHVM.  Can you remove failing steps from build which doesn't meet minimal requirements for php by silex(two with php 5.5 and one with HHVM)?

jdreesen commented:
> Now build is failing for php 5.5.x version which isn't compatible with silex 2. Silex to requires php in >=5.5.9.

Travis uses `PHP 5.5.38` which Silex is compatible with, so that's not the problem. Albeit I'm not sure why [`Job #83.1`](https://travis-ci.org/PHP-DI/Silex-Bridge/jobs/304646318) failed, though (the tests run successfully on my machine).

> Other test is with HHVM. Can you remove failing steps from build which doesn't meet minimal requirements for php by silex(two with php 5.5 and one with HHVM)?

As @mnapoli said: feel free to drop HVVM from Travis.
The PHP 5.5 tests should not be dropped because Silex supports these PHP versions, so we should too.


[`Job #83.5`](https://travis-ci.org/PHP-DI/Silex-Bridge/jobs/304646322) failed because it installed the lowest possible dependencies that are allowed in the `composer.json`, which is `Symfony 2.8`,  but this version does not support the `ArgumentResolverInterface` yet (seems it was added in `3.1`).
I think there is still a bit of work to do here, to support Symfony 2.8/3.0 and Symfony >= 3.1.

jdreesen commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 /vendor/
 /composer.phar
 /composer.lock
+/.idea/
```
This should be removed here and added to the global `.gitignore` file on your computer instead, because it has nothing to do with this library (but with the IDE you are using).

For more information on global `.gitignore`, see:
- https://help.github.com/articles/ignoring-files/#create-a-global-gitignore
- http://augustl.com/blog/2009/global_gitignores/
- https://github.com/github/gitignore/tree/master/Global#globally-useful-gitignores

kordos replied:
Ok, I can remove this. It's just easier to click in IDE and mark as ignored :D

jdreesen replied:
It is, until you put that line into your global .gitignore. Because from then on it is ignored for every (even new) project on your computer ;)

kordos replied:
That also depends on particular IDE which may not recognize global git settings. But no problem I will remove it.

kordos commented:
> Travis uses PHP 5.5.38 which Silex is compatible with, so that's not the problem. Albeit I'm not sure why Job #83.1 failed, though (the tests run successfully on my machine).

Nope, minimal version for silex 2 is 5.5.9 which is a bit higher than 5.5.38

> As @mnapoli said: feel free to drop HVVM from Travis.
> The PHP 5.5 tests should not be dropped because Silex supports these PHP versions, so we should too.

I'm not familiar with Travis. Can you do it?

jdreesen commented:
>     Travis uses PHP 5.5.38 which Silex is compatible with, so that's not the problem. Albeit I'm not sure why Job #83.1 failed, though (the tests run successfully on my machine).
> 
> Nope, minimal version for silex 2 is 5.5.9 which is a bit higher than 5.5.38

The minimum PHP version for Silex is 5.5.9, so there should be no problem with PHP 5.5.38 because it is much newer ;)

>     As @mnapoli said: feel free to drop HVVM from Travis.
>     The PHP 5.5 tests should not be dropped because Silex supports these PHP versions, so we should too.
> 
> I'm not familiar with Travis. Can you do it?

Yes, I can do it (I'll try to find some time this evening).

kordos commented:
> The minimum PHP version for Silex is 5.5.9, so there should be no problem with PHP 5.5.38 because it is much newer ;)

Right, It should work.

jdreesen committed:
Add PHP 7.1 to and drop HHVM from the test matrix

jdreesen committed:
Use local PHPUnit to run the tests on Travis

jdreesen committed:
Update minimum PHPUnit version to latest 4.x release and allow PHPUnit 5

jdreesen commented:
I removed HHVM from the Travis test matrix and added PHP 7.1.
I also required newer PHPUnit versions and changed Travis to use the local version instead of the global one (the global version somehow failed on PHP 5.5).

The last job with the lowest possible dependencies still fails, because we use interfaces that don't exist in Symfony < 3.1 at the moment (see comments above).

kordos commented:
> I removed HHVM from the Travis test matrix and added PHP 7.1.
I also required newer PHPUnit versions and changed Travis to use the local version instead of the global one (the global version somehow failed on PHP 5.5).

great. thanks :D

> The last job with the lowest possible dependencies still fails, because we use interfaces that don't exist in Symfony < 3.1 at the moment (see comments above).

Ok I will fix that

Commit:
ControllerResolver - bring back previous version for symfony 2.8

Commit:
updated tests for symfony 2.8

Commit:
Merge remote-tracking branch 'origin/silex-2' into silex-2

Commit:
updated tests for symfony 2.8

kordos commented:
I add some changes for symfony 2.8.x and now it should work also with that version. Build also works

kordos commented:
Hi. Builds are passing now and It works also with symfony 2.8+. Can you merge this pull request?

Commit:
HttpKernelServiceProvider - removed request definition

kordos commented:
Hi, I resolved all your comments. Can you merge it?

kordos commented:
@mnapoli, @jdreesen can you merge it?

mnapoli commented:
:+1:, busy times lately :)
